Oauth, local login, & logout

bethqiang opened the pull request:
Closes #9 & #14

parky22 committed:
add auth files

parky22 committed:
added edits from bones oauth updates

parky22 committed:
changed transcend.env.json file name

parky22 committed:
add cookie-session npm; set up some FE

parky22 committed:
update oauth branch with master

parky22 committed:
update .env file to format with dotenv

parky22 committed:
add users routes from bones

parky22 committed:
some changes made with Omri

bethqiang committed:
Merge master

bethqiang committed:
Fix db issue + housekeeping

Server is at least running normally now, I think

bethqiang committed:
FE login component fires action on submit

bethqiang committed:
Add local login via passport

bethqiang committed:
Add logout functionality

Currently logout button is in Login component, will have to find a better place for it

bethqiang committed:
Add Google OAuth

Also set state for user after OAuth

bethqiang committed:
Clean up, clean up, everybody do your share

Clean up, clean up, everybody everywhere

bethqiang committed:
Login looks purty

bethqiang committed:
Add redirect after local login

bethqiang committed:
Favicon got lost. It's found now!

(Also, move favicon to public. Seemed slightly better.)

bethqiang committed:
Move login div down slightly

bethqiang committed:
Clean up

JoeyDarbyshire commented on browser/app.html:
```diff
@@ -4,6 +4,10 @@
     <title>Transcend</title>
     <meta name="description" content="A social and educational VR environment.">
+    <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
+    <link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-wvfXpqpZZVQGK6TAh5PVlGOfQNHSoD2xbE+QkPxCAFlNEevoEH3Sl0sibVcOQVnN" crossorigin="anonymous">
```
Should we just npm install this instead? 
https://www.npmjs.com/package/font-awesome

bethqiang replied:
Probably.

JoeyDarbyshire commented on browser/react/components/Login.js:
```diff
@@ -0,0 +1,81 @@
+    );
+  }
+
+  onLoginSubmit (event) {
```
Shouldn't this just be passed as a prop from MapDispatchToProps?  Then you wouldn't need to bind it I believe.

bethqiang replied:
Perhaps. Will look into it.

JoeyDarbyshire commented on server/api.js:
```diff
@@ -0,0 +1,8 @@
+
+api.use('/auth', require('./auth'));
+
+// No routes matched? 404.
```
All one route it could match on...... lmao

bethqiang replied:
Deferring to @parky22 on this one. I personally like having the option of being able to throw in more routes super easily if/when the time comes...

JoeyDarbyshire replied:
I do as well, I was just making a joke xD

bethqiang replied:
Apparently my 1am brain doesn't know how to do jokes or sarcasm lol.

JoeyDarbyshire commented on server/auth.js:
```diff
@@ -0,0 +1,102 @@
+      where: {
+        googleId: profile.id
+      },
+      defaults: info
```
nice

bethqiang replied:
I can't tell if this is sarcasm or not...?

JoeyDarbyshire replied:
definitely not, I like the defaults tag.  hadn't seen it before for findOrCreate

JoeyDarbyshire commented on server/auth.js:
```diff
@@ -0,0 +1,102 @@
+);
+
+// Send user info to frontend
+auth.get('/whoami', (req, res) => res.send(req.user));
```
This could either send back the user, or do a redirect to /login like you have above on line 90

bethqiang replied:
Future problems, since no authorization infrastructure is in place yet. (I'm assuming this is referring to authorization and not authentication.)

JoeyDarbyshire commented on browser/react/components/Login.js:
```diff
@@ -0,0 +1,81 @@
+    document.body.style.background = 'url(/images/background.png) no-repeat top center fixed';
+    document.body.style.backgroundSize = 'cover';
+  }
+
```
I feel like all of this could be done in CSS instead.

bethqiang replied:
It can...if you want every single view (e.g., our home page, and each of our rooms, etc.) to have the same background.

JoeyDarbyshire replied:
couldn't the background image just be for a div that's 100% height and width... give the div an id and only render it on the login page

bethqiang replied:
As per Slack discussion/exchanging of screenshots, will keep mount/unmount.

bushidocodes replied:
I've added a class name to the scenes equal to the href suffix. e.g. the <a-scene> returned by 'localhost:1337/beth/' has a class of 'beth'. Sounds like you won't need it, but it'll be there when my stuff gets merged in.

JoeyDarbyshire commented on browser/stylesheets/login.css:
```diff
@@ -0,0 +1,136 @@
+.form-control,
+.form-control:focus,
+.login-button,
+.login-with-google {
```
Why are these all on separate lines?

bethqiang replied:
Because I think it looks much cleaner than having them all on the same line.

JoeyDarbyshire commented on browser/react/components/Login.js:
```diff
@@ -0,0 +1,81 @@
+/* ----------------- COMPONENT ------------------ */
+
+class Login extends React.Component {
+  constructor (props) {
```
all of my following comments are basically trying to make this a dumb component.  I like dumb components.  I think they look cleaner.  I'm a big redux fan personally.

bethqiang replied:
I am too, but I'm not sure that's possible with the necessary mounting/unmounting that needs to be done.

bethqiang committed:
Changes from Joey's code review

- Move loginSubmit functionality to login function
- Serve font awesome from npm install instead of cdn

bushidocodes commented on browser/redux/reducers/auth.js:
```diff
@@ -0,0 +1,50 @@
+export const login = (username, password) => {
+  return dispatch =>
+    axios.post('/api/auth/local/login', { username, password })
+    .then(() => browserHistory.push('/'))
```
I think the behavior here should be that an end user should be able to hit any URL on our domain, get redirected to /login if not logged in, and upon successful login, get redirected to the URL that the user originally entered. This would allow for bookmarking of VR rooms, etc.

parky22 replied:
yess authorization is the next step.

bushidocodes commented on browser/redux/reducers/auth.js:
```diff
@@ -0,0 +1,50 @@
+    .catch(err => console.log(err.message));
+};
+
+export const logout = () =>
```
Perhaps we should add a 3d link for logout in the VR world to allow it to be clicked when in headset.

parky22 replied:
maybe it could be another one of your orbs.

bushidocodes commented on server/auth.js:
```diff
@@ -0,0 +1,102 @@
+// Local login
+auth.post('/local/login', (req, res, next) => {
+  passport.authenticate('local', {
+    successRedirect: '/'
```
Ditto on the earlier comment. I think that we should redirect to that URL that the user entered before getting redirected to the login page.

bushidocodes reviewed (approved):
All of my comments regard future refinements that would should consider adding as issues. Good work!
